### PR TITLE
Implement maximum length for wxTextCtrl on wxQt

### PR DESCRIPTION
**Pull request by oneeyeman1**
The code in this PR was tested on the Qt version 5.7.
It would be nice to check this code under the latest Qt version (5.12).

Or maybe the 5.12 implementation is useless?

Let me know what is the preference here.

Thank you.

### CODE REVIEW

File: include/wx/qt/textctrl.h [hunk 1]
```diff
@@ -73,7 +73,8 @@ class WXDLLIMPEXP_CORE wxTextCtrl : public wxTextCtrlBase
 private:
     QLineEdit *m_qtLineEdit;
     QTextEdit *m_qtTextEdit;
-
+    virtual QLineEdit *GetEditable() const wxOVERRIDE;
```
Comment (vadz) [inline include/wx/qt/textctrl.h]: All the other wx styles declare the functions _before_ the member variables, not after. Could you please do it like this here too for consistency?

File: include/wx/qt/textentry.h [hunk 2]
```diff
@@ -36,13 +38,15 @@ class WXDLLIMPEXP_CORE wxTextEntry : public wxTextEntryBase
     virtual bool IsEditable() const;
     virtual void SetEditable(bool editable);
     
+    virtual void SetMaxLength(unsigned long len);
```
Comment (vadz) [inline include/wx/qt/textentry.h]: Extremely minor, but a blank line before `protected:` would be welcome.

File: include/wx/qt/textentry.h [hunk 3]
```diff
@@ -36,13 +38,15 @@ class WXDLLIMPEXP_CORE wxTextEntry : public wxTextEntryBase
...
 protected:
     virtual wxString DoGetValue() const;
     virtual void DoSetValue(const wxString& value, int flags=0);
 
     virtual wxWindow *GetEditableWindow();
 
 private:
+    virtual QLineEdit *GetEditable()  const { return NULL; };
```
Comment (vadz) [inline include/wx/qt/textentry.h]: It doesn't make sense to call this function `GetEditable()`, it's called like this in wxGTK because it returns `GtkEditable`, but here it really should be called `GetLineEdit()` or maybe `QTGetLineEdit()`.

Also, shouldn't it be pure virtual?

File: src/qt/textctrl.cpp [hunk 4]
```diff
@@ -21,10 +21,14 @@ class wxQtLineEdit : public wxQtEventSignalHandler< QLineEdit, wxTextCtrl >
 {
 public:
     wxQtLineEdit( wxWindow *parent, wxTextCtrl *handler );
-
+protected:
+    virtual void keyPressEvent(QKeyEvent *event) wxOVERRIDE;
 private:
     void textChanged(const QString &text);
     void returnPressed();
+#if QT_VERSION >= QT_VERSION_CHECK( 5, 12, 0)
```
Comment (vadz) [inline src/qt/textctrl.cpp]: This condition is repeated several times and is really not very readable. Could you please define `QT_HAS_INPUT_REJECTED` (or something making even more sense if you can think of a better name) preprocessor symbol once, using this expression, and then test for it instead?

File: src/qt/textctrl.cpp [hunk 5]
```diff
@@ -61,6 +69,36 @@ void wxQtLineEdit::returnPressed()
     }
 }
 
+#if QT_VERSION >= QT_VERSION_CHECK( 5, 12, 0)
+void wxQtLineEdit::inputRejected()
+{
+    wxTextCtrl *handler = GetHandler();
+    if( handler )
```
Comment (vadz) [inline src/qt/textctrl.cpp]: Could you please insert a space between `if` and `(`? Here and elsewhere?

File: src/qt/textctrl.cpp [hunk 6]
```diff
@@ -61,6 +69,36 @@ void wxQtLineEdit::returnPressed()
...
+    {
+        QLineEdit *line_edit = ((QLineEdit *) handler->GetHandle());
+        if( line_edit )
+        {
+            int max = line_edit->maxLength();
```
Comment (vadz) [inline src/qt/textctrl.cpp]: This temporary variable seems unnecessary.

File: src/qt/textctrl.cpp [hunk 7]
```diff
@@ -61,6 +69,36 @@ void wxQtLineEdit::returnPressed()
     }
 }
 
+#if QT_VERSION >= QT_VERSION_CHECK( 5, 12, 0)
+void wxQtLineEdit::inputRejected()
+{
+    wxTextCtrl *handler = GetHandler();
+    if( handler )
+    {
+        QLineEdit *line_edit = ((QLineEdit *) handler->GetHandle());
```
Comment (vadz) [inline src/qt/textctrl.cpp]: Is this different from what `GetEditable()` returns? If so, a comment explaining why/how would be useful. If not, why not just use it here?

File: src/qt/textentry.cpp [hunk 8]
```diff
@@ -7,8 +7,11 @@
 
 // For compilers that support precompilation, includes "wx.h".
 #include "wx/wxprec.h"
-
+#include "wx/textctrl.h"
```
Comment (vadz) [inline src/qt/textentry.cpp]: `wxTextEntry` really shouldn't depend on `wxTextCtrl`, why is this needed here?

**Review by vadz — CHANGES_REQUESTED** [review]
Sorry, but unfortunately there are several questions about/problems with this change.

**Comment by oneeyeman1** [comment]
@vadz ,
Almost all those comments are style ones. I will fix those and re-submit.
However, I'd like to know if we even want to check for 5.12 here.

Could you answer please?

**Comment by vadz** [comment]
I don't know. Actually, I've forgotten to ask about this: what's the limitation of the other implementation and why/how is 5.12+ version better?

**Comment by oneeyeman1** [comment]
@vadz ,
According to the Qt docs, there is now inputRejected signal, which can be caught and we can send our event from it. And since it is specifically designed to check this stuff I think it should be better than catching the key press signal and verifying the length of the string in the text control every time.

OTOH, I don't know the implementation details. However, I hope that this signal is sent from the textEdited/textChanged signals and not from the key press one.

Problem is this signal is new in 5.12.

**Comment by vadz** [comment]
OK, we're _definitely_ not going to check in any code which uses a version of Qt that you haven't tested at all. I'm sorry, but I really don't understand why this could possible be a good idea.

Anyhow, please make sure to test with the version of Qt you do have now and document any limitations you encounter.

**Comment by oneeyeman1** [comment]
@vadz ,
OK.
But the original question remains - should the implementation be changed in the future utilizing the sugnal introduced in 5.12?
If not I will remove the conditions and leave the current implementation only?

**Comment by vadz** [comment]
Again, I don't know. But please do remove the untested code. If you think there is an advantage in adding it later, we can do it. But we're not going to add some code that literally nobody, including its author, knows what is it for!

**Comment by oneeyeman1** [comment]
@vadz ,
I think I fixed everything on this PR.
I completely removed the code related to 5.12+ as this is a brand new release and it will take some time to accept it to most distros.
Now, entering the text from the keyboard and overflowing the control works. Also pressing Ctrl+V works as well.

Do you want this to be as a side note in the documentation?

Thank you.

**Comment by vadz** [comment]
> Do you want this to be as a side note in the documentation (as a comment to SetMaxLength() or to the event)?

Sorry, what do you mean by "this"?

File: include/wx/qt/textctrl.h [hunk 9]
```diff
@@ -71,9 +71,11 @@ class WXDLLIMPEXP_CORE wxTextCtrl : public wxTextCtrlBase
     virtual QScrollArea *QtGetScrollBarsContainer() const;
 
 private:
+    virtual QLineEdit *QTGetLineEdit() const wxOVERRIDE;
+
     QLineEdit *m_qtLineEdit;
     QTextEdit *m_qtTextEdit;
-
+
```
Comment (vadz) [inline include/wx/qt/textctrl.h]:
```suggestion

```

File: include/wx/qt/textentry.h [hunk 10]
```diff
@@ -36,13 +38,16 @@ class WXDLLIMPEXP_CORE wxTextEntry : public wxTextEntryBase
...
+    virtual void SetMaxLength(unsigned long len) wxOVERRIDE;
+
 protected:
     virtual wxString DoGetValue() const;
     virtual void DoSetValue(const wxString& value, int flags=0);
 
     virtual wxWindow *GetEditableWindow();
 
 private:
+    virtual QLineEdit *QTGetLineEdit()  const { return NULL; };
```
Comment (vadz) [inline include/wx/qt/textentry.h]:
```suggestion
    virtual QLineEdit *QTGetLineEdit()  const { return NULL; }
```

File: src/qt/textctrl.cpp [hunk 11]
```diff
@@ -61,6 +62,16 @@ void wxQtLineEdit::returnPressed()
     }
 }
 
+void wxQtLineEdit::keyPressEvent(QKeyEvent *event)
+{
+    wxTextCtrl *handler = GetHandler();
+    if ( !event->text().isEmpty() && maxLength() == handler->GetValue().length() )
```
Comment (vadz) [inline src/qt/textctrl.cpp]: Could you please add a comment explaining why is the first part of the test needed?

Also, this:
```suggestion
    if ( !event->text().isEmpty() && maxLength() == handler->GetLastPosition() )
```
would be more efficient, which might matter here.

File: src/qt/textentry.cpp [hunk 12]
```diff
@@ -106,3 +108,9 @@ wxWindow *wxTextEntry::GetEditableWindow()
     return NULL;
 }
 
+void wxTextEntry::SetMaxLength(unsigned long len)
+{
+    QLineEdit *editor = QTGetLineEdit();
+    if( editor )
```
Comment (vadz) [inline src/qt/textentry.cpp]:
```suggestion
    if ( editor )
```

File: src/qt/textentry.cpp [hunk 13]
```diff
@@ -7,8 +7,10 @@
 
 // For compilers that support precompilation, includes "wx.h".
 #include "wx/wxprec.h"
-
```
Comment (vadz) [inline src/qt/textentry.cpp]: Could you please avoid making unrelated whitespace changes? TIA!

**Review by vadz — COMMENTED** [review]
Sorry, still a couple of question, as well as the apparently unavoidable style nitpicks.

**Comment by oneeyeman1** [comment]
@vadz 
>>  Do you want this to be as a side note in the documentation (as a comment to SetMaxLength() or to the event)?

> Sorry, what do you mean by "this"?

"Now, entering the text from the keyboard and overflowing the control works. Also pressing Ctrl+V works as well.".

**Comment by vadz** [comment]
No, we don't usually document that a method works as intended.

**Review by catalinr — CHANGES_REQUESTED** [review]
`wxTextEntry` is not supposed to know about the various controls that implement its functionality.
What will happen if Qt people decide to support maximum length for multi-line controls? Will there be a `QTGetMultiLineEdit` too? And another one for combobox?
`SetMaxLength` is already virtual. It can be implemented directly in `wxTextCtrl` to simply forward the limit to the underlying Qt control.
The docs should be updated to mention that only single line text controls support this feature under Qt.

The current attempt for emitting `wxEVT_TEXT_MAXLEN` is not complete (and IMO it's very unlikely it could ever be). What happens when the line is "full" and you right click with the mouse and choose `Paste`, or drag and drop some more text, or simply press `Ctrl` (without 'V'!)? Or if the line is _not_ "full" but you attempt to paste text that would exceed the quota?
For this I'd suggest not applying any code changes, but rather changing the docs to mention that `wxEVT_TEXT_MAXLEN` is not emitted under Qt, at least for now.

**Comment by oneeyeman1** [comment]
> No, we don't usually document that a method works as intended.

The current implementation checks for entering the text from keyboard and also checks for Ctrl+V. In the latter case if the paste is bigger than the maxlen the text is accepted to satisfy the maxlen requirements and no maxlen event is emitted. Every consecutive Ctrl+V attempt will emit the event.

I don't know if any of the supported platforms works like this.

**Comment by oneeyeman1** [comment]
@catalinr ,

> wxTextEntry is not supposed to know about the various controls that implement its functionality.
> What will happen if Qt people decide to support maximum length for multi-line controls? Will there be > a QTGetMultiLineEdit too? And another one for combobox?
> SetMaxLength is already virtual. It can be implemented directly in wxTextCtrl to simply forward the > limit to the underlying Qt control.
> The docs should be updated to mention that only single line text controls support this feature under Qt.

I think that this can be done because the docs already mentions this for GTK+
Also, there is a patch on trac that implements this for GTK+ multiline text control. It could be adapted to support Qt as well.

> The current attempt for emitting wxEVT_TEXT_MAXLEN is not complete (and IMO it's very unlikely it > could ever be). What happens when the line is "full" and you right click with the mouse and choose 
> Paste, or drag and drop some more text, or simply press Ctrl (without 'V'!)? Or if the line is not "full" 
> but you attempt to paste text that would exceed the quota?

I tried to explain it in the reply to Vadim comment at least partially.

> For this I'd suggest not applying any code changes, but rather changing the docs to mention that 
> wxEVT_TEXT_MAXLEN is not emitted under Qt, at least for now.

It probably will be possible for the QT version >= 5.12, but not all *nices are up with it yet, as the release happens very recently.

File: src/qt/textctrl.cpp (hunk 11 above)
Comment (oneeyeman1) [reply src/qt/textctrl.cpp]: @vadz,

GetLastPosition() is not yet implemented for wxTextCtrl under Qt, and so for now I think I will use the current code

**Comment by catalinr** [comment]
> > What happens when the line is "full" and you right click with the mouse and choose
> > Paste, or drag and drop some more text, or simply press Ctrl (without 'V'!)? Or if the line is not "full"
> > but you attempt to paste text that would exceed the quota?
> 
> I tried to explain it in the reply to Vadim comment at least partially.

Not really. See the question before the last one.
And about that, the event is supposed to be emitted even when only a part of the [pasted | dropped | inserted] text is accepted. Implementing a functionality automatically implies that the requirements are known in advance, doesn't it?

> We can also apply the patch and document the limitation about the paste and drag'n'drop.

Those are not the only scenarios that wouldn't work. What about programmatically appending text? What about keys that do not generate characters, like modifiers or their combinations? Ctrl+**C**? Maybe other cases too

> At least we will have something that works.

That would be a false claim and the end result will be bugs that complain about missing or wrong functionality.

Why don't you just add the simplest implementation for setting the limit and then, separately, concentrate on sending the event, if you still feel like doing it?

**Comment by vadz** [comment]
FWIW I agree that it would be nice to start with the minimal and uncontroversial change implementing `SetMaxLength()`. The rest could follow later.

Could you please update this PR (or maybe create a new one, whichever is simpler) to do it?

**Comment by vadz** [comment]
I don't know what has happened here, but the original branch seems to have been deleted and the patch doesn't apply at all (i.e. it wasn't made against any version of code in master), so it's not really useful to keep this opened.

I'd also like to note that `wxTextEntry` is not really implemented in wxQt currently and the first thing to do would be to provide its implementation, as it seems like it shouldn't be difficult to do by just adding `QtGetLineEdit()` private abstract method and implementing it in `wxTextCtrl` and `wxComboBox`.
